### PR TITLE
Feature fuzz null mode

### PR DESCRIPTION
**Pull request by cheatguy**

### CODE REVIEW

File: fuzz/tls_init/CMakeLists.txt [hunk 1]
```diff
@@ -0,0 +1,74 @@
...
+
+if(NOT SGX)
+    #set(CMAKE_C_FLAGS "-fPIE ${CMAKE_C_FLAGS}")
+    set(CMAKE_C_FLAGS ${CMAKE_C_FLAGS})
+    set(CMAKE_CXX_COMPILER "/usr/bin/clang++")
+    set(CMAKE_CXX_FLAGS "-g -fsanitize=address,fuzzer  ${CMAKE_CXX_FLAGS}")
+    set(CMAKE_EXE_LINKER_FLAGS "${CMAKE_EXE_LINKER_FLAGS}")
+endif()
+
+if(SGX)
```
Comment (haosanzi) [inline fuzz/tls_init/CMakeLists.txt:11]: Since pr is about host mode, please delete sgx mode

File: fuzz/tls_init/fuzz_init.cc [hunk 2]
```diff
@@ -0,0 +1,79 @@
...
+    strcpy(conf.crypto_type, "nullcrypto");
+    
+    conf.cert_algo = RATS_TLS_CERT_ALGO_DEFAULT;
+    conf.flags = RATS_TLS_CONF_FLAGS_MUTUAL;
+
+    FuzzedDataProvider fuzzed_data(data + sizeof(conf), size - sizeof(conf));
+    claim_t custom_claims[CUSTOM_CLAIMS_SIZE];
+    std::vector<std::string> str_lists;
+    for(int i=0;i<CUSTOM_CLAIMS_SIZE;i++){
+        //const char * str = fuzzed_data.ConsumeBytesWithTerminator(50,'\0').data();
```
Comment (haosanzi) [inline fuzz/tls_init/fuzz_init.cc:45]: please refer to [link](https://github.com/inclavare-containers/inclavare-containers/blob/master/CONTRIBUTING.md#format-c-codes) to format your code.

File: fuzz/tls_init/fuzz_init.cc [hunk 3]
```diff
@@ -0,0 +1,79 @@
...
+        /*这里不能使用上面的方法,否则会有空悬指针的问题*/
+        std::vector<char> vec_str = fuzzed_data.ConsumeBytesWithTerminator(50,'\0');
```
Comment (haosanzi) [inline fuzz/tls_init/fuzz_init.cc:47]: Please change Convert Chinese to English

File: fuzz/tls_init/fuzz_init.cc [hunk 4]
```diff
@@ -0,0 +1,79 @@
...
+#include <stdint.h>
+#include <stddef.h>
+#include <string.h>
+#include "rats-tls/api.h"
+#include "rats-tls/log.h"
+#include "rats-tls/claim.h"
+#include "internal/core.h"
+}
+#include <fuzzer/FuzzedDataProvider.h>
+#include <vector>
```
Comment (haosanzi) [inline fuzz/tls_init/fuzz_init.cc:16]: Does the `libfuzzer-dev` package need to be installed? Please explain in the readme.

File: fuzz/tls_init/fuzz_init.cc [hunk 5]
```diff
@@ -0,0 +1,79 @@
...
+    }
+    memcpy(&conf, data, sizeof(conf));
+    conf.log_level  = RATS_TLS_LOG_LEVEL_DEFAULT;
+    conf.api_version = 0;
+    
+    /*fuzz log level*/
+    /*fuzz round could not be too huge, that leads to unexpected log_level*/
+    strcpy(conf.attester_type, "nullattester");
+    strcpy(conf.verifier_type, "nullverifier");
+    strcpy(conf.tls_type, "nulltls");
```
Comment (haosanzi) [inline fuzz/tls_init/fuzz_init.cc:35]: Does other fileds in rats_tls_conf_t such as `attester_type ` or `verifier_type ` etc need to be fuzzed?

File: fuzz/tls_init/fuzz_init.cc (hunk 4 above)
Comment (haosanzi) [reply fuzz/tls_init/fuzz_init.cc:16]: Do we need to add a fuzz compilation option? When the flag is on, the fuzz folder is compiled.

Comment (cheatguy) [reply fuzz/tls_init/fuzz_init.cc:16]: install the new version of clang, libfuzzer is included in it.

Comment (cheatguy) [reply fuzz/tls_init/fuzz_init.cc:16]: > Do we need to add a fuzz compilation option? When the flag is on, the fuzz folder is compiled.

if needed, I would change the top CMakeLists, and add a -DFUZZ for this. now I just use add_subdictionary() to build fuzz in any condition.

File: fuzz/tls_init/fuzz_init.cc (hunk 5 above)
Comment (cheatguy) [reply fuzz/tls_init/fuzz_init.cc:35]: You mean fill them totally  with random string or different type like "sgx_la" .
And for host mode, I also should use different attester/verifier type too?

Comment (haosanzi) [reply fuzz/tls_init/fuzz_init.cc]: yes. In host mode, there are different attester/verifier type. We cannot assume that we are using the null type, the user may output different types, and the program will output different error logs.

**Comment by haosanzi** [comment]
replaced by #183

**Pull request closed without merging**